modules/openapi-generator/src/main/resources/Java/libraries/aws-java-client/ApiClient.mustache

```diff
@@ -0,0 +1,593 @@
+        this.credentials = new BasicAWSCredentials(accessKey, secretKey);
+    }
+
+    /**
+    * Set region for a signed call to API
+    *
+    * @param region The region of the API being invoked
+    */
+    public void setAWSRegion(String region) {
+        this.region = region;
+    }
+
+    /**
+    * Set service for a signed call to API
+    *
+    * @param service The region of the API being invoked
+    */
+    public void setService(String service) {
+        this.service = service;
+    }
+
+    /**
+    * Set path for a signed call to API
+    *
+    * @param path The region of the API being invoked
+    */
+    public void setPath(String path) {
+        this.path = path;
+    }
+
+    /**
+    * Set method for a signed call to API
+    *
+    * @param method The region of the API being invoked
+    */
+    public void setMethod(String method) {
+        this.method = method;
+    }
+
+    /**
+    * Set queryParams for a signed call to API
+    *
+    * @param queryParams The region of the API being invoked
+    */
+    public void setQueryParams(List<Pair> queryParams) {
+        this.queryParams = queryParams;
+    }
+
+    /**
+    * Set collectionQueryParams for a signed call to API
+    *
+    * @param collectionQueryParams The region of the API being invoked
+    */
+    public void setCollectionQueryParams(List<Pair> collectionQueryParams) {
+        this.collectionQueryParams = collectionQueryParams;
+    }
+
+    private class SimpleAwsErrorHandler implements HttpResponseHandler<SdkBaseException> {
+        @Override
+        public AmazonClientException handle(HttpResponse response) throws Exception {
+            String toThrow = response.getStatusCode() + " " + response.getStatusText() + "\n" + response.getHeaders();
+            AmazonClientException e = new AmazonClientException(toThrow);
+            return e;
+        }
+        @Override
+        public boolean needsConnectionLeftOpen() {
+            return false;
+        }
+    }
+
+    /**
+    * Build HTTP call with the given options.
+    *
+    * @param path The sub-path of the HTTP URL
+    * @param method The request method, one of "GET", "HEAD", "OPTIONS", "POST", "PUT", "PATCH" and "DELETE"
+    * @param queryParam The query parameters
+    * @param collectionQueryParams The collection query parameters
+    * @return The HTTP call
+    * @throws ApiException If fail to serialize the request body object
+    */
+    public Response buildCall(String path, String method, List<Pair> queryParam, List<Pair> collectionQueryParams) throws ApiException {
+        Request apiRequest = buildRequest(path, method, queryParam, collectionQueryParams);
+        return newCall(apiRequest);
+    }
+
+    /**
+    * Makes call to API and returns the Response
+    *
+    * @param apiRequest The request to the API
+    * @return The Response from the API
+    */
+    public Response newCall(Request apiRequest) {
```

Comment:
In regards to this entire class... I'm confused when I review the public interface. I can't understand if/why you need so many 1) public methods and 2) public setters. 

Please review the public interface for your new classes - you should ALWAYS prefer the strictest level of visibility unless there's a reason not to (i.e. private by default). If you want something testable, sometimes we use package private to allow the test access when it's defined in the same package location.

As for setters - they make it very very difficult to reason through the lifecycle and maintenance of an instance of the class.

It's a classic issue... How do I build this Foo object? Oh I didn't realize I had to call setter X/Y/Z to make this object perform method doThing()... It's confusing and error-prone.

If you absolutely need them....? OK I suppose - maybe explain why in the class or per-method javadoc?

Reply:
I have removed some of the unnecessary public methods and setters. Many of them were leftover from how the default client is generated with OpenAPI generator. Unsure why there was a need for so many things to be public, but I was just following the pattern. I changed all of the functions I implemented to private except the ones that the defaultClient needs to interact with to make the call. 

The other functions that were already written and still needed to keep everything functioning properly have been left alone. I removed a few of the setters to allow for the user to invoke a new buildCall function from main if they choose to overwrite the parameters needed for the signing process. This can be invoked by retrieving the api from the defaultClient and calling buildCallWithParams which takes the user provided credentials and region.